test/jetstream_test.go

```diff
@@ -7976,26 +7990,33 @@ func TestJetStreamAckExplicitMsgRemoval(t *testing.T) {
 				DeliverSubject: sub2.Subject,
 				FilterSubject:  "foo.bar",
 				AckPolicy:      server.AckExplicit,
+				AckWait:        25 * time.Millisecond,
 			})
 			if err != nil {
 				t.Fatalf("Unexpected error adding consumer: %v", err)
 			}
 			defer o2.Delete()
 
-			checkFor(t, time.Second, 100*time.Millisecond, func() error {
-				if nmsgs, _, _ := sub2.Pending(); nmsgs != toSend {
-					return fmt.Errorf("Did not receive correct number of messages: %d vs %d", nmsgs, toSend)
-				}
-				return nil
-			})
-
 			// Those messages should be redelivered to the 2nd consumer
-			for i := 0; i < toSend; i++ {
+			for i := 1; i <= toSend; i++ {
 				m, err := sub2.NextMsg(time.Second)
 				if err != nil {
-					t.Fatalf("Error acking message: %v", err)
+					t.Fatalf("Error receiving message %d: %v", i, err)
 				}
 				m.Respond(nil)
+
+				sseq, dseq, dc, _ := o2.ReplyInfo(m.Reply)
```

Comment:
For all the following checks: if I add the artificial delay in stream.go:
```
		if !hasInterest {
			time.Sleep(100 * time.Millisecond)
			mset.store.RemoveMsg(seq)
		}
```
then these checks fail. What I mean is that is these checks not making too many assumptions on the timing of things? Is that going to be a flapper?
Would dseq not be possibly 3 and 4 (and not 4 and 5) if say the unsubscribe() is "fully processed" by the time new messages are sent. Adding a 250ms delay between unsubscribe and sending messages 3 and 4 show that it is the case.

To sum-up, I am not sure that this test should check the delivery/redelivery sequence but simply that message 3 and 4 are received (in any order, but not several times the same) .

Reply:
Let me take a look and simulate your delay as well and see.

Reply:
If you don't put the delay before RemoveMsg() but instead pause (like 250ms) between sub2.Unsubscribe() and sending messages 3 and 4, you will likely have message 3 and 4 delivered "normally". So the point of my comments was that these checks may make the test flap.

Reply:
ok agree will remove check on consumer sequence.